caffe2/contrib/tensorrt/tensorrt_op_trt.cc

```diff
@@ -0,0 +1,224 @@
+#include "caffe2/contrib/tensorrt/tensorrt_op_trt.h"
+#include "caffe2/core/logging.h"
+
+#include <numeric>
+#include <unordered_map>
+
+namespace caffe2 {
+
+namespace {
+// Note that input of trt tensor is in CHW format, while our tensor is NCHW
+// \return -1 if there is dimension mismatch between C2 tensor and trt tensor.
+// Otherwise, return the multiplicaton of CHW dimensions
+int64_t CheckDims(
+    const nvinfer1::Dims& nv_dims,
+    const std::vector<TIndex>& c2_dims) {
+  if (nv_dims.nbDims + 1 != c2_dims.size()) {
+    return -1;
+  }
+  int64_t chw = 1;
+  for (int i = 0; i < nv_dims.nbDims; ++i) {
+    if (nv_dims.d[i] != c2_dims[i + 1]) {
+      return -1;
+    }
+    chw *= nv_dims.d[i];
+  }
+  return chw;
+}
+
+} // namespace
+
+// Upon construction, we build the inference enigne by deserializing from
+// protobuf string. And since we know the input/output blobs, we can do the
+// binding here too.
+TensorRTOp::TensorRTOp(const OperatorDef& operator_def, Workspace* ws)
+    : Operator<CUDAContext>(operator_def, ws),
+      logger_((nvinfer1::ILogger::Severity)(
+          OperatorBase::GetSingleArgument<int>("log_verbosity", 2))),
+      max_batch_size_(OperatorBase::GetSingleArgument<int>("max_batch_size", 1)) {
+  {
+    auto engine_string =
+        OperatorBase::GetSingleArgument<std::string>("serialized_engine", "");
+    CAFFE_ENFORCE(!engine_string.empty(), "Empty serialized TensorRT engine!");
+    auto trt_runtime = InferObject(nvinfer1::createInferRuntime(logger_));
+    // TODO(support trt plugin factory)
+    trt_engine_ = InferObject(trt_runtime->deserializeCudaEngine(
+        engine_string.data(), engine_string.size(), nullptr));
+  }
```

Comment:
Need a way to specify target device for inference (use `operator_def->device_option().cuda_gpu_id()`)

Reply:
From the sample code, looks like it's set via plain `cudaSetDevice`. Maybe we can push/pop the old device id if it's not the same as `operator_def->device_option().cuda_gpu_id()` ?

Reply:
isn't it regular cudaSetDevice? I actually don't know how TRT exposes devices

Reply:
Yes, it's the regular `CudaSetDevice`.